### PR TITLE
[WIP] qt5 update to 5.14.0

### PR DESCRIPTION
**Pull request by Johnnynator**
# Tested on
- [x] x86_64 (compiled and runtime)
- [x] x86_64-musl (compiled and runtime)
- [ ] i686 ()
- [ ] armv7l-musl
- [ ] armv7l 
- [ ] armv6l-musl
- [ ] armv6l (compiled without webengine)
- [ ] aarch64-musl (compiled)
- [ ] aarch64
- [ ] ppc64le
- [ ] ppc64le-musl
- [ ] ppc64
- [ ] ppc64-musl
- [ ] ppc
- [ ] ppc-musl
# TODO:

- fix cross (for some qmake packages)
- rebase webengine ppc patches @q66 
- `-no-feature-qdoc` instead of custom `qttools/src/src.pro` ?
- fix webengine loading resources from `/share` instead of `/usr/share`

# Bugs

- webengine currently tries to fetch resources from `/share` instead of `/usr/share
- Plasmashell wayland on musl crashes regulary

### CODE REVIEW

**Comment by PaperMountainStudio** [comment]
qt5 and qt5-doc have the same source, it should be better to merge the templates

**Comment by Johnnynator** [comment]
> qt5 and qt5-doc have the same source, it should be better to merge the templates

I'm aware of that. But it is this way on purpose. And now that I have split out webengine we wouldn't be able to just integrate it, without not having the webengine docs build. But I should add `qt5-base-docs` to the deps of `qt5-doc`

**Comment by CameronNemo** [comment]
you are timing out hard.

**Comment by streaksu** [comment]
Any news about this? Void could make good use of the granularity this patch proposes and the new qt version.

**Comment by Johnnynator** [comment]
> Any news about this? Void could make good use of the granularity this patch proposes and the new qt version.

I currently don't have much time to get all the (broken) cross compilation stuff fixed. QMake breaks like every release in that regard for us.

**Comment by travankor** [comment]
FWIW Alpine is on qt5.14.1 and their build files look relatively more simple so I wonder how they are cross-compiling 🤔

**Comment by Johnnynator** [comment]
> FWIW Alpine is on qt5.14.1 and their build files look relatively more simple so I wonder how they are cross-compiling thinking

They aren't cross compiling usually.

**Comment by mintsuki** [comment]
Is there an ETA for this version bump considering it's been now more than 5 months since this PR was opened?

Qt is quite an important package and, at least its `qt5-webengine` component, should see fairly constant maintainance as quite a few browsers (such as `qutebrowser`, `falkon`, and `otter`) depend on it for, amongst other things, CVE patches.

I have been waiting a long time for this PR to be completed but at this point I am starting to lose hope and perhaps I should give it a shot myself, even though I am not superfamiliar with `xbps`.

**Comment by q66** [comment]
I don't think we're exactly in a hurry, 5.13 is not yet EOL, still officially supported until June, so if some major CVE arises, there should be a patch release. But yes, would be nice to finally get it updated.

You're obviously free to help, if you want to speed it up.

**Comment by Johnnynator** [comment]
> I have been waiting a long time for this PR to be completed but at this point I am starting to lose hope and perhaps I should give it a shot myself, even though I am not superfamiliar with `xbps`.

You don't need to know much about xbps, just a lot about qmake and how it likes to break for our cross use cases. I really hope that Qt6 changing to CMake will make this less problematic.

> Is there an ETA for this version bump considering it's been now more than 5 months since this PR was opened?

Sry, can't give any ETA.

**Comment by travankor** [comment]
>You don't need to know much about xbps, just a lot about qmake and how it likes to break for our cross use cases. I really hope that Qt6 changing to CMake will make this less problematic.

+1 (and I hope they get rid of the MOC)

>Sry, can't give any ETA.

QT 5.15 will be the last one in the 5.X series, so that might be the one to focus on, instead of QT 5.14.

**Comment by pullmoll** [comment]
Just FYI I also have a branch where I've been trying to get 5.14.1 built:
https://github.com/pullmoll/void-packages/tree/qt5-5.14
This branch builds qt5 for native architectures. Cross building is where there's a lot of work left to do and for which I don't currently find sufficient time.

**Comment by xtraeme** [comment]
well, I'm starting to think that we should move on and continue if something is not cross buildable.

**Comment by xtraeme** [comment]
because, cross compilation is really an intensive hard task. So we should push changes and when we've got cross buildable push it.

Otherwise we will suffer with the same issues that plagued pkgsrc...

**Comment by Johnnynator** [comment]
> Just FYI I also have a branch where I've been trying to get 5.14.1 built:
> https://github.com/pullmoll/void-packages/tree/qt5-5.14
> This branch builds qt5 for native architectures. Cross building is where there's a lot of work left to do and for which I don't currently find sufficient time.

Non cross also works here :)



> well, I'm starting to think that we should move on and continue if something is not cross buildable.

I agree on that one for less used software, but Qt is quite popular and I'd argue that people with e.g. a Pinebook would like to have Qt applications in the repo. We should just get a native aarch64 builder...

**Comment by pullmoll** [comment]
I think it would be worth a try to go modular with qt5. That would mean build qt5-base native and later cross build qt5-base using the qmake build style and host tools. I'm not sure if we can have different build styles for native and cross builds in one package, though. It should be possible AFAICT.

If it works this would avoid having to build the host tools on the fly when cross compiling and the big amount of patching required to also get the target arch tools built in the same go.

In other words: us trying to keep the monolithic build approach is also causing much of our trouble.

Once we'd have the native qt5-base (and qt5-qmake, qt5-host-tools etc.) all of the qt5 modules should then be built using the qmake build style and its generated `qt.conf` for the target, native or cross. The *only* thing required for this to work is the qmake build style producing good results for all cases.

**Comment by q66** [comment]
this PR kind of started modularizing some of these things, I think it's too much work to get it done within the constraints though, we need to get the update out and then we can worry about changing things further :)

**Comment by Chocimier** [comment]
@pullmoll: build_style is regular variable that can be set under `if $cross_build`, no problem here. It is utilized in cmake template.

@xtraeme: Please note that when package do not cross compile, anything depending can't be updated or rebuild for cross-build archs too.

**Comment by mintsuki** [comment]
I attempted this update myself and sent a PR here: https://github.com/void-linux/void-packages/pull/20296

I compiled it and tested it for x86_64 glibc only. I'll test x86_64-musl shortly and report back.

**Comment by travankor** [comment]
>I compiled it and tested it for x86_64 glibc only. I'll test x86_64-musl shortly and report back.

The main issue is that ARM/PPC builds of QT5 are not build-able with Void's infrastructure (hence, why the problem is cross-compiling with qmake). Alpine Linux gets around this by having native build machines for each architecture.

>Please note that when package do not cross compile, anything depending can't be updated or rebuild for cross-build archs too.

I feel like the buildbot replacement should handle this automatically.

**Comment by q66** [comment]
@travankor only ARM, void doesn't do official ppc packages and my ppc repos are completely native. Nevertheless, having qt be cross-compilable is crucial, to any supported target.

**Comment by travankor** [comment]
>only ARM, void doesn't do official ppc packages and my ppc repos are completely native.

I thought this was only a temporary arrangement? Well, it's a shame if void won't accept PPC as on official arch.

>Nevertheless, having qt be cross-compilable is crucial, to any supported target.

Well, seems like this will be fixed for QT6
https://wiki.qt.io/CMake_Port

**Comment by Johnnynator** [comment]
Don't hold your breath that CMake will make it easier. It is only better documented and usually easier to configure but has even worse cross compile capabilities than qmake.

**Comment by q66** [comment]
@travankor if it becomes an official architecture, it will be native as well.
